ArduPlane/navigation.cpp

```diff
@@ -152,19 +156,47 @@ void Plane::calc_airspeed_errors()
 
 void Plane::calc_gndspeed_undershoot()
 {
- 	// Use the component of ground speed in the forward direction
-	// This prevents flyaway if wind takes plane backwards
-    if (gps.status() >= AP_GPS::GPS_OK_FIX_2D) {
-	    Vector2f gndVel = ahrs.groundspeed_vector();
-		const Matrix3f &rotMat = ahrs.get_rotation_body_to_ned();
-		Vector2f yawVect = Vector2f(rotMat.a.x,rotMat.b.x);
-        if (!yawVect.is_zero()) {
-            yawVect.normalize();
-            float gndSpdFwd = yawVect * gndVel;
-            groundspeed_undershoot = (aparm.min_gndspeed_cm > 0) ? (aparm.min_gndspeed_cm - gndSpdFwd*100) : 0;
-        }
+
+    // Use the component of ground speed in the forward direction
+    // This prevents flyaway if wind takes plane backwards
+    const Matrix3f &rotMat = ahrs.get_rotation_body_to_ned();
+    Vector2f yawVect = Vector2f(rotMat.a.x,rotMat.b.x);
+
+    // Use integrator based gndspd_undershoot controller
+    if (gps.status() >= AP_GPS::GPS_OK_FIX_2D &&
+        is_flying() && !yawVect.is_zero() &&
+        (aparm.min_gndspeed_cm > 0)) {
+
+        float nav_bearing = radians(0.01f*nav_controller->target_bearing_cd());
+        Vector2f navVect = Vector2f(cosf(nav_bearing), sinf(nav_bearing));
+        navVect.normalize();
+
+        Vector2f gndSpdVect = ahrs.groundspeed_vector();
+        yawVect.normalize();
+
+        // Using the component of ground speed in the forward direction only
+        // prevents flyaway if wind takes plane backwards, but may
+        // result in underestimating the actual ground speed along track.
+        // Hence, use the max of both estimates.
+
+        float gndspd_forward     = yawVect * gndSpdVect;
+        float gndspd_along_track = navVect * gndSpdVect;
+        float gndSpdFwd = MAX(gndspd_forward, gndspd_along_track);
+
+        float gndspd_undershoot_error = aparm.min_gndspeed_cm - gndSpdFwd*100;
+        // compensate for the lag in airspeed controller (coverting to cm/sec)
+        gndspd_undershoot_error -= airspeed_error*100;
+
+        const float dt = 0.1f;  // this is running at 10Hz
+        const float K_I = 0.2f; // fixed integration gain
```

Comment:
how do you feel about making this a function of TCONST?

Reply:
hm, this part is called from a function "update_GPS_10Hz()", but if you prefer having it more robust against future change in timing than I could redo it to something similar to the dt from update_fbwb_speed_height(). The thing is, I can't find any TCONST that would be defined within the Plane class, though there are a couple definitions of provate parameters like PTCH2SRV_TCONST or TECS_LAND_TCONST, but I guess that is not what you meant. So what do you suggest?

Reply:
I meant making the blending adjustable via TECS_TIME_CONST and TECS_LAND_TCONST. I'm thinking of the scenario where someone is doing airspeed demand changes quickly during a reverse-thrust landing. You go through approach to pre-flare to flare in just a matter of a couple seconds so some people would prefer a step input (or at least a very fast one) to the speed demand so that it reacts as fast as possible